go/runtime/txpool/txpool.go

```diff
@@ -231,38 +234,46 @@ func (t *txPool) SubmitTxNoWait(ctx context.Context, tx []byte, meta *Transactio
 }
 
 func (t *txPool) submitTx(ctx context.Context, rawTx []byte, meta *TransactionMeta, notifyCh chan *protocol.CheckTxResult) error {
-	tx := newTransaction(rawTx, txStatusPendingCheck)
+	// todo: in the new design, we'll be submitting to the check queue first, which won't have "first seen time"
+	// metadata. seems like it could still work out, as checking is single threaded, so we can still preserve the
+	// original order when we finish checking and offer to the main queue
 
+	// todo: can we go without this?
 	// Skip recently seen transactions.
-	if _, seen := t.seenCache.Peek(tx.hash); seen {
-		t.logger.Debug("ignoring already seen transaction", "tx_hash", tx.hash)
+	// todo: hash used to be cached
+	h := hash.NewFromBytes(rawTx)
+	if _, seen := t.seenCache.Peek(h); seen {
+		t.logger.Debug("ignoring already seen transaction", "tx_hash", h)
 		return fmt.Errorf("duplicate transaction")
 	}
 
 	// Queue transaction for checks.
 	pct := &PendingCheckTransaction{
-		Transaction: tx,
-		notifyCh:    notifyCh,
-	}
-	if meta.Local {
-		pct.flags |= txCheckLocal
+		tx:       rawTx,
+		notifyCh: notifyCh,
 	}
 	if meta.Discard {
-		pct.flags |= txCheckDiscard
+		pct.dstQueue = nil
+	} else if meta.Local {
+		pct.dstQueue = t.localQueue
+	} else {
+		pct.dstQueue = t.mainQueue
 	}
 
 	return t.addToCheckQueue(pct)
 }
 
 func (t *txPool) addToCheckQueue(pct *PendingCheckTransaction) error {
+	// todo: hash used to be cached
```

Comment:
Yeah it would be great if this was the case to avoid recomputing the tx hashes all over the place.

Reply:
let's see how many uses remain after this draft is more complete. gonna be sad if it's only for logging. but still

Reply:
ok we're back to caching this